lib/experimental/results.nim

```diff
@@ -0,0 +1,187 @@
+##[
+This module implements a result-type which encapsulates both a success and a
+failure in one value. Only one of the two states can be active.
+
+Based on https://github.com/disruptek/badresults, which was itself based on
+https://github.com/arnetheduck/nim-result.
+
+Exception translation
+=====================
+
+Using `get` or `value` raises an exception if the `Result` stores a failure.
+For `Result[T, E]` the exception raised is:
+ * if `E` is a `ref` type that inherits from `Exception`, the `error` value
+ * if a custom overload of `toException <#toException, E>`_ exists, the result
+  of `toException(r)`
+ * otherwise, a new `ResultError[E]` with `error` as the payload
+
+]##
+
+type
+  ResultError*[E] = ref object of ValueError
+    error: E
+
+  Result*[T, E] = object
+    case o: bool
+    of false:
+      e: E
+    of true:
+      v: T
+
+proc error*[T, E](self: Result[T, E]): lent E {.inline.} =
+  ## Retrieves the error from a `Result`; raises `ResultError[void]`
+  ## if the `Result` is not in error.
+  if self.isErr:
+    result = self.e
+  else:
+    raise ResultError[void](msg: "Result does not contain an error")
+
+proc unsafeGet*[T, E](self: var Result[T, E]): var T {.inline.} =
+  ## Fetches the value of a `Result` if set, undefined behavior if unset.
+  ##
+  ## See also:
+  ## * `unsafeGet <options.html#unsafeGet,Option[T]>`_
+  assert not isErr(self)
+  result = self.v
+
+proc unsafeGet*[T, E](self: Result[T, E]): lent T {.inline.} =
+  ## Fetches the value of a `Result` if set, undefined behavior if unset.
+  ## See also:
+  ## * `unsafeGet <options.html#unsafeGet,Option[T]>`_
+  assert not isErr(self)
+  result = self.v
+
+proc unsafeGet*[E](self: Result[void, E]) {.inline.} =
+  ## Does nothing if `Result` is set, undefined behavior if unset.
+  ## See also:
+  ## * `unsafeGet <options.html#unsafeGet,Option[T]>`_
+  assert not self.isErr
+
+func newResultError[E](e: sink E; s: sink string): ResultError[E] {.inline, nimcall.} =
+  ## capturing ResultError...
+  ResultError[E](error: e, msg: s)
+
+template toException*[E](err: E): ResultError[E] =
+  ## Refer to `this <#exception-translation>`_ for more information
+  mixin `$`
+  when compiles($err):
+    newResultError(err, "Result isErr: " & $err)
+  else:
+    newResultError(err, "Result isErr; no `$` in scope.")
+
+template raiseResultError[T, E](self: Result[T, E]) =
+  mixin toException
+  when E is ref Exception:
+    if self.error.isNil: # for example Result.default()!
+      raise ResultError[void](msg: "Result isErr; no exception.")
+    else:
+      raise self.error
+  else:
+    raise self.error.toException
+
+proc ok*[E](R: typedesc[Result[void, E]]): Result[void, E] =
```

Comment:
These ok procs feel far too benignly named for how easily they can break logic. They're outright nuking a result, for someone reading code using Result they'd rightly be shocked as to the severity of the mutation.

Reply:
How about `initSuccess`?

Reply:
Sounds good to me!